process: expose uv_rusage on process.resourceUsage()

vmarchaud opened the pull request:
As discussed in https://github.com/nodejs/diagnostics/issues/161,
the core should expose important metrics about the runtime, this PR's
goal is to let user get the number of io request made, and lower level
metrics like the page faults and context switches.

~~The `make -j4 test` are not passing locally because of a failing test `test/parallel/test-worker-prof.js` but i believe it's not related to my changes.~~
Concerning the tests, i didn't really tested the output of the function mainly because it might make the test flaky depending on the platform and runtime behavior.

This is my first PR on the core, so i might have forgotten some requirements.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+    * `inblock` {integer}
+    * `oublock` {integer}
+    * `nvcsw` {integer}
+    * `nivcsw` {integer}
```
I’m wondering if we should spell out these names fully in camelCase, e.g. `involuntaryContextSwitches`? I think the other diagnostic APIs on `process` might set enough precedent for that?

vmarchaud replied:
I actually started with full names but i removed them later on, mainly because i believe it's easier for user to check what they each one actually means by checking against the kernel documentation. Otherwise we should document the mapping between both linux names and ours, what do you think ?

addaleax replied:
I think using full names and documenting the mapping makes sense, but I’d also like to hear what other people think about it too.

mscdex replied:
Seeing as this is mostly a *nix thing (only a couple or so properties are supported on Windows), I think it makes more sense to keep the original names used by the kernel.

bnoordhuis replied:
I'm a bit torn on this but I think I side with @addaleax because their names are cryptic even by unix standards.

vmarchaud replied:
So what should we do ? I believe more people are in favor of using full names.

BridgeAR replied:
:+1: for using full names.

addaleax commented on src/node_process_methods.cc:
```diff
@@ -278,6 +278,29 @@ void GetActiveHandles(const FunctionCallbackInfo<Value>& args) {
+  int err = uv_getrusage(&rusage);
+  if (err) {
+    // On error, return the strerror version of the error code.
+    Local<String> errmsg = OneByteString(args.GetIsolate(), uv_strerror(err));
```
Can we use `Environment::ThrowUVException` instead of stringifying the error and returning it? See e.g. the implementation of `Cwd()` in this file for how to use that. In that case, we should also be able to drop the new error code here (which would imo be a good thing, because error codes should reflect the type of error that happened, and not its origin).

vmarchaud replied:
:+1:

Trott commented on test/parallel/test-resources-usage.js:
```diff
@@ -0,0 +1,32 @@
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Generally, new test files should not have the copyright boilerplate.

vmarchaud replied:
:+1:

Trott commented:
> This is my first PR on the core, so i might have forgotten some requirements.

Looks like you didn't forget anything. Thanks for the pull request and welcome!

Trott commented on test/parallel/test-resources-usage.js:
```diff
@@ -0,0 +1,12 @@
+// Copyright Joyent, Inc. and other Node contributors.
```
You can remove the copyright entirely!

vmarchaud replied:
My bad i wasn't scrolling to the top of the file ahah

cjihrig replied:
I've unresolved this conversation because a copyright line is still present.

Trott commented:
Any reason this would make `test-worker-prof` less reliable? It's failing a lot in CI here, but I'm not sure if it's because of the changes here or something else?

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
 
+## process.resourcesUsage()
+<!-- YAML
+added: v12.x.x
```
```suggestion
added: REPLACEME
```

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
 
 Additional documentation is available in the [report documentation][].
 
+## process.resourcesUsage()
```
Can you name this `resourceUsage()` (resource instead of resources) throughout the PR.

vmarchaud replied:
Wouldn't it make more sense to keep `resources` since there are metrics related to different resources (io, cpu and memory)?

addaleax replied:
@vmarchaud I think it’s just that noun+noun compounds in English generally use the singular form for the all but the last word/head of the phrase. So, if you really wanted to emphasize the fact that this function provides multiple sets of information, you could use `resourceUsages()`, but I’d prefer Colin’s suggestion and just go with the singular. And if nothing else, consistency with `memoryUsage()` is valuable, too.

vmarchaud replied:
Oh i see, english is not my main english so i didn't knew about that.

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+    * `nvcsw` {integer}
+    * `nivcsw` {integer}
+
+The `process.resourcesUsage()` method return the current resources usage
```
```suggestion
The `process.resourcesUsage()` method returns the resource usage
```

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+added: v12.x.x
+-->
+
+* Returns: {Object}
```
How did you decide what to expose and what to omit from http://docs.libuv.org/en/v1.x/misc.html#c.uv_rusage_t

vmarchaud replied:
I removed heap/cpu related fields since they are exposed throught `process.cpuUsage()` and `process.memoryUsage()`. Other fields i removed are supported on neither linux or windows (see https://linux.die.net/man/2/getrusage description for each field)

bnoordhuis replied:
I have a mild preference for exposing the fields as-is, supported or not, duplicating other APIs or not.

If that obsoletes `process.cpuUsage()`, then so be it. You could reimplement it as a JS-only shim around `process.resourceUsage()`.

(`process.memoryUsage()` still has merit because it also reports the V8 heap size.)

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+-->
+
+* Returns: {Object}
+    * `minflt` {integer}
```
I'm not sure if it matters at the JS level, but these are all unsigned integers.

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+for the current process.
+To understand what these value mean, please refer to
+[getrusage(2)](http://man7.org/linux/man-pages/man2/getrusage.2.html)
+On Windows platform, only `inblock` and `oublock` are supported,
```
```suggestion
On Windows platforms, only `inblock` and `oublock` are supported,
```

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+To understand what these value mean, please refer to
+[getrusage(2)](http://man7.org/linux/man-pages/man2/getrusage.2.html)
+On Windows platform, only `inblock` and `oublock` are supported,
+all other fields will report zero as value.
```
```suggestion
with all other fields containing zero.
```

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+On Windows platform, only `inblock` and `oublock` are supported,
+all other fields will report zero as value.
+
+For example, the code:
```
I think you can drop this line completely.

cjihrig commented on doc/api/process.md:
````diff
@@ -1819,6 +1819,46 @@ process.report.writeReport();
+
+<!-- eslint-skip -->
+```js
+{
````
This can be combined with the `console.log()` example, with the output commented out.

cjihrig commented on test/parallel/test-resources-usage.js:
```diff
@@ -0,0 +1,12 @@
+const assert = require('assert');
+
+const r = process.resourcesUsage();
+assert.ok(typeof r.minflt === 'number');
```
It might be useful to also assert that the values are `>= 0`.

vmarchaud replied:
As said on the PR, i believe it will render the test flaky. Since the test is run without doing anything else, there are generally no memory page swap and almost no IO request, but it can change from platform to platform i suppose.

addaleax replied:
@vmarchaud These values are all unsigned when coming from libuv – so they have to be `>= 0`.

vmarchaud replied:
My bad i read `> 0` on @cjihrig message, will correct !

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,41 @@ process.report.writeReport();
+The `process.resourceUsage()` method returns the resource usage
+for the current process.
+To understand what these value mean, please refer to
+[getrusage(2)](http://man7.org/linux/man-pages/man2/getrusage.2.html)
```
Sorry, I missed this before, but I believe the linked URL should be at the bottom of the file with the other links.

cjihrig commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,41 @@ process.report.writeReport();
+for the current process.
+To understand what these value mean, please refer to
+[getrusage(2)](http://man7.org/linux/man-pages/man2/getrusage.2.html)
+On Windows platforms, only `inblock` and `oublock` are supported,
```
Just noting that this sentence will need to be updated pending two decisions in this PR:

- Abbreviated or expanded field names.
- Inclusion of all the fields present in libuv.

cjihrig commented on test/parallel/test-resource-usage.js:
```diff
@@ -0,0 +1,11 @@
+const assert = require('assert');
+
+const r = process.resourceUsage();
+assert.ok(r.minflt >= 0);
```
I'd leave the previous `typeof` assertions in place as well, just to be safe. For example, if by some future mistake these values became numeric strings, these assertions wouldn't catch it.

vmarchaud commented:
I've switched to full names, added the mapping between node names and libuv ones. Also exposed all the fields even if not supported as @bnoordhuis said.

BridgeAR commented:
What benefit do we have by exposing entries that are not supported on any platform?

addaleax commented:
@bridgear Future-proofing and standards conformance would be the two big pro arguments, I imagine?

addaleax commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,91 @@ process.report.writeReport();
+    * `unsharedStackSize` {integer}
+    * `minorPageFault` {integer}
+    * `majorPageFault` {integer}
+    * `swapedOut` {integer}
```
I think this should be `swappedOut`?

addaleax commented on doc/api/process.md:
```diff
@@ -1819,6 +1819,91 @@ process.report.writeReport();
+for the current process.
+All of these values come from the `uv_getrusage` call which returns
+[this struct][uv_rusage_t], here the mapping between node and libuv:
+- `userCPUTime` map to `ru_utime` computed in microseconds.
```
I think `map to` should be `maps to` everywhere?

vmarchaud commented:
I believe the PR is good to be merged expect if someone has a blocker on the exposed fields ?

jasnell commented on src/node_process_methods.cc:
```diff
@@ -278,6 +278,38 @@ void GetActiveHandles(const FunctionCallbackInfo<Value>& args) {
+  fields[12] = rusage.ru_msgrcv;
+  fields[13] = rusage.ru_nsignals;
+  fields[14] = rusage.ru_nvcsw;
+  fields[15] = rusage.ru_nivcsw;
```
nit: this would be a bit more readable/maintainable if these were named constants instead of just literal index numbers

vmarchaud replied:
Not sure how to do that exactly, i never really coded in C++

jasnell commented on test/parallel/test-resource-usage.js:
```diff
@@ -0,0 +1,27 @@
+assert(typeof rusage.voluntaryContextSwitches === 'number' &&
+  rusage.voluntaryContextSwitches >= 0);
+assert(typeof rusage.involuntaryContextSwitches === 'number' &&
+  rusage.involuntaryContextSwitches >= 0);
```
The following might make it a bit more maintainable if items are removed/added

```js
[
  'userCPUTime',
  'systemCPUTime',
  'maxRSS',
  'sharedMemorySize',
  'unsharedDataSize',
  'unsharedStackSize',
  'minorPageFault',
  'majorPageFault',
  'swappedOut',
  'fsRead',
  'fsWrite',
  'ipcSent',
  'ipcReceived',
  'signalsCount',
  'voluntaryContextSwitches',
  'involuntaryContextSwitches'
].forEach((n) => {
  assert.strictEqual(typeof rusage[n], 'number');
  assert(rusage[n] >= 0);
});
```

Trott commented on doc/api/process.md:
```diff
@@ -2335,3 +2420,7 @@ cases:
+[uv_rusage_t]: http://docs.libuv.org/en/v1.x/misc.html#c.uv_rusage_t
+[wikipedia_minor_fault]: https://en.wikipedia.org/wiki/Page_fault#Minor
+[wikipedia_major_fault]: https://en.wikipedia.org/wiki/Page_fault#Major
+[process.cpuUsage]: #process_process_cpuusage_previousvalue
```
Nit: Can you move this up a few lines to keep all the link-text in ASCII order?

vmarchaud commented:
I've rebased on master hoping that the test on worker-prof will pass now

vmarchaud commented:
I've checked the test failures and it doesn't seems to be related to this PR, what are the next steps in this case ?

vmarchaud commented:
@Trott I'm not sure what are the next steps, do you think the PR can be merged ?

addaleax commented:
@vmarchaud Thanks for the ping!

Landed in 6271ef005eb311c819b7ab041e09f50e4155ec1e :tada:
